### ISSUE TITLE
[Enhancement]: API: Search customers by first_name and last_name

### ISSUE THREAD

**Issue by joshbetz** [woothemes/woocommerce#33383]
### Describe the solution you'd like

Allow customers to be searched by first name and last name. Since we use WP_User_Query, we can search by the [default fields that WordPress uses](https://developer.wordpress.org/reference/classes/wp_user_query/#search-parameters) (ID, user_login, user_nicename, user_email, user_url) and we also add display_name.

Note: The first_name and last_name fields are stored in usermeta, so it could be expensive to query them.

**Comment by coreymckrill** [comment]
To elaborate a bit more on the current state of things:

`wc/v3/customers?search=foo`

* The API controller adds wildcards to both sides of the string to make it a partial string match.
* When API instantiates the `WP_User_Query`, it doesn't specify `search_columns`, so WordPress has some limited logic to try and determine which column would be the best fit, but defaults to `array( 'user_login', 'user_url', 'user_email', 'user_nicename', 'display_name' )` if the search string isn't obviously an email address or URL.
* As already mentioned, `first_name` and `last_name` are usermeta, so they aren't included in the query. Technically `display_name` could contain first name, last name, a combination of the two, or something else entirely, but it appears that when a new user is created, it defaults to first name only, and customers generally don't have a UI for changing this.
* It turns out that the customers data store actually has [a method](https://github.com/woocommerce/woocommerce/blob/0c5f93bb39170d3a8d419f7d87e969e386f31588/plugins/woocommerce/includes/data-stores/class-wc-customer-data-store.php#L508-L571) for searching first/last name along with the other fields. It accomplishes this by actually running two separate queries and then combining the results. But, it's not compatible with the API endpoint because it can't accommodate other query parameters like `role`, `include`, `exclude`, `order`, `orderby`, etc.

`wc/v3/customers?email=foo`

* This searches only the `user_email` column. But it's full match only, not partial match. In fact, `email=foo` will return an error because it's not a valid whole email address.
* If a request contained both `search` and `email`, the `email` parameter would override the other.

`wc-analytics/reports/customers?search=foo&searchby=name`

* This will concatenate the `first_name` and `last_name` fields and then do a partial match on them. This is the closest to what I believe the desired behavior actually is.
* Since this is in the context of analytics, the data objects returned have different fields than `wc/v3/customers`. Probably most significantly, they don't have separate, detailed billing and shipping info.
* As described [here](https://github.com/woocommerce/woocommerce-ios/pull/7746), this endpoint can be used to work around the name search limitation by first doing a partial match name search that returns user IDs, and then requesting the data objects for those specific IDs from `wc/v3/customers`.

**Comment by coreymckrill** [comment]
Some possible solutions:

* Modify the behavior of the current `search` parameter so that it does partial match on first/last name along with the other default fields. In order to maintain backcompat, this alternative behavior could be toggled on using a separate boolean request parameter, something like `search_first_last=true`.
* Add a new request parameter, perhaps `name`, that would work a bit like the existing `email` parameter. Except it would be partial match and search on first/last name along with the other default fields.
* Add a new endpoint, `wc/v3/customers/search`, that would take a `query` parameter that would then be used to do a partial match.

**Comment by kidinov** [comment]
@coreymckrill 

To confirm, that the current state of things is still the same as you described in this [message](https://github.com/woocommerce/woocommerce/issues/33383#issuecomment-1287550556)?

**Comment by coreymckrill** [comment]
Correct, the current state of things hasn't changed. Currently discussing which of the possible solutions mentioned above would be the best to pursue further.

**Comment by m0untaingoat** [comment]
Hey, just wondering if there's been any movement on this? I have a customer that is asking about a possible related issue. 

zd 7417578

**Comment by StreamlinedStartup** [comment]
This still hasn't been addressed?

**Comment by Malaiac** [comment]
That is a common problem amongst commercants : searching for customers in the WooCommerce / Customers is the obvious way, but it ends up much less powerful than the WordPress / Users menu

**Comment by samrat-biswas-git** [comment]
Hi @joshbetz - thanks for the report.

I know this has been open for a very long time, and I appreciate your patience. We're currently going through some older WooCommerce issues, identifying what still needs fixing, and tidying up what's already fixed or out of scope.

I've taken a look and marked this ready for the engineering team to pick up.

What I'm seeing: Current customer collection queries on trunk still only search core user fields like login, email, nicename, URL, and display name. WooCommerce's customer data store also still has a separate `search_customers()` helper that explicitly queries `first_name` and `last_name`, which confirms the gap is real and still unresolved for the REST customer collection endpoints.
